docs: add air-gapped ck8s environment docs (#782)
elastisys/compliantkubernetes: docs/operator-manual/air-gapped-environment.md

# CK8s in Air-gapped Environments

!!!warning

>     Each air-gapped environment is different. Therefore, this page should only be taken as a guideline, not as "works out-of-the-box instructions". Contact [Elastisys Support](https://elastisys.com/self-managed/) if you need help.

Reviewer: We tend to recommend to put each sentence in its own line, but we don't enforce it uniformly. @aarnq and @salehsedghpour please voice your opinion on whether this fix is a "nice to have" or a "must have".

Reviewer 2: I'd say it would be preferred, and _hopefully_ I'll manage to find a linter setup that works with our combo of plugins sometime soon. :sweat_smile:

## Background

> In an air-gapped environment, machines are isolated from insecure networks such as the public Internet. Air-gapping is used for environments that handle highly confidential data such as military or governmental systems, or in life-critical systems, for example, in nuclear power plants or for medical equipment. This document will provide with instructions on how to configure Compliantkubernetes Apps to work inside an air-gapped environment.

Reviewer: ```suggestion
In an air-gapped environment, machines are isolated from insecure networks such as the public Internet. Air-gapping is used for environments that handle highly confidential data such as military or governmental systems, or in life-critical systems, for example, in nuclear power plants or for medical equipment. This document provides guidelines on how to configure Compliantkubernetes Apps to work inside an air-gapped environment.
```

Let's avoid the word "instructions", since we just said we don't provide instructions. :smile:

## System Context Diagram

> The following is a generic system context diagram over an air-gapped CK8s environment. The diagram shows the Platform Administrator Machine that has access to both the Internet and the air-gapped environment. Inside the air-gapped environment, the Compliantkubernetes clusters can access other services residing on the same private network such as an object storage for long time storage, a container registry, DNS and NTP servers, etc.

Reviewer: ```suggestion
The following is a generic system context diagram over an air-gapped CK8s environment. The diagram shows the Platform Administrator Machine that has access to both the Internet and the air-gapped environment. Inside the air-gapped environment, the clusters forming a Compliant Kubernetes environment can access other services residing on the same private network such as an object storage for long-term storage, a container registry, DNS and NTP servers, etc.
```

![Diagram](../img/air-gapped-real.drawio.png)

!!!note

>     \* For Certificate Provisioning, we could open ports to allow DNS01 or HTTP01 challenges to a public certificate authority, such as [LetsEncrypt](https://letsencrypt.org/). However, if you want the environment to be truly air-gapped, then you need to figure out an air-gapped solution for PKI.

Reviewer: ```suggestion
    \* For Certificate Provisioning, you may want to allow DNS01 or HTTP01 challenges to a public certificate authority, such as [LetsEncrypt](https://letsencrypt.org/). However, if you want the environment to be truly air-gapped, then you need to figure out an air-gapped solution for PKI.
```

> ## Configuring Air-gapped Compliantkubernetes Apps

Reviewer: ```suggestion
## Configuring Air-gapped Compliant Kubernetes Apps
```

> These instructions will show how to configure Compliantkubernetes Apps to work in an air-gapped environment. For setting up the Kubernetes layer, please refer to the [Kubespray air-gap installation documentation](https://kubespray.io/#/docs/offline-environment) or [Cluster-API documentation](https://cluster-api.sigs.k8s.io/clusterctl/configuration.html?highlight=release#image-overrides) for further instructions.

Reviewer: ```suggestion
These guidelines will show how to configure Compliant Kubernetes Apps to work in an air-gapped environment. For setting up the Kubernetes layer, please refer to the [Kubespray air-gap installation documentation](https://kubespray.io/#/docs/offline-environment) or [Cluster-API documentation](https://cluster-api.sigs.k8s.io/clusterctl/configuration.html?highlight=release#image-overrides) for further instructions.
```

This guide will assume that we have a Platform Administrator Machine that can access both the Internet and the air-gapped environment over SSH, this includes SSH access to both cluster nodes and hosts for the different offline services described in the system context diagram above. This "Platform Administrator Machine" can be your local machine or a bastion host.

> Start by initializing your Compliantkubernetes Apps config (see the quickstart section [here](https://github.com/elastisys/compliantkubernetes-apps?tab=readme-ov-file#quickstart) for more info on initializing and deploying Compliantkubernetes Apps) with the `air-gapped` flavor by setting the `CK8S_FLAVOR` environment variable:

Reviewer: ```suggestion
Start by initializing your Compliant Kubernetes Apps config (see the quickstart section [here](https://github.com/elastisys/compliantkubernetes-apps?tab=readme-ov-file#quickstart) for more info on initializing and deploying Compliant Kubernetes Apps) with the `air-gapped` flavor by setting the `CK8S_FLAVOR` environment variable:
```

```bash
export CK8S_FLAVOR=air-gapped
```

### Container Images and Registry Mirroring

> We need that all container images used for the Kubernetes layer and for Compliantkubernetes Apps are available in the air-gapped environment. This can be done by setting up a private container registry (e.g. `Harbor`) that can act as a [registry mirror or registry cache](https://docs.docker.com/docker-hub/mirror/). But first these images need to be added to the private container registry. The following will provide some commands that can be used to get all of the images used by Compliantkubernetes Apps.

Reviewer: Please make the following changes in this paragraph:

- use "you" for the reader, "we" for the writer;
- prefer active sentences, e.g., "The following commands can be used" -> "Use the following commands to get ..."
- please spell "Compliant Kubernetes"

#### Migrating Images From a Running Environment with Script

> There is [a useful script](https://kubespray.io/#/contrib/offline/README?id=manage-offline-container-imagessh) in the upstream Kubespray repository that can be used to get all images from an already running Compliantkubernetes cluster by simply pointing the `KUBECONFIG` environment variable to the correct cluster(s). If possible it is recommended to use this script for migrating images to a local registry.

Reviewer: ```suggestion
There is [a useful script](https://kubespray.io/#/contrib/offline/README?id=manage-offline-container-imagessh) in the upstream Kubespray repository that can be used to get all images from an already running Compliant Kubernetes environment by simply pointing the `KUBECONFIG` environment variable to the correct cluster(s). If possible it is recommended to use this script for migrating images to a local registry.
```

I think my suggestion conforms better to our glossary, but please double-check: https://elastisys.io/compliantkubernetes/glossary/

Also, you seem to use the word "local registry" here and "Container Registry" in the air-gapped system context diagram. Please use "container registry mirror" in both.

We can then copy over the images to the host running the container registry, after which the images can be tagged and pushed to. To copy over images, you can store them as tarballs, e.g. if you use Docker on your local machine:

```bash
RUNTIME=nerdctl # see note for docker and podman
REGISTRY= # set this to the public registry containing the image, e.g. ghcr.io
IMAGE_NAME= # include repository, image name and tag, e.g. aquasecurity/trivy-operator:0.17.1
IMAGE_FILE_NAME=image_file.tar
sudo ${RUNTIME} pull ${REGISTRY}/${IMAGE_NAME}
sudo ${RUNTIME} save -o ${IMAGE_FILE_NAME}.tar ${REGISTRY}/${IMAGE_NAME}
```

!!!note

    Be wary of images with digests as tags, as when saving the container image when using the `docker` or `podman` runtime and then loading it, the image will lose its digest (see issue discussing this [here](https://github.com/moby/moby/issues/22011)). If you use this approach, you may need to override the digest or image tag used in the Compliantkubernetes Apps config. Using a tool like `nerdctl` will however preserve the correct image digest.

!!!tip

>     If your Platform Administrator Machine has access to the air-gapped container registry, you can use a tool like [`regctl` or `regsync`](https://github.com/regclient/regclient) which can help syncing images between repositories. This way, images will keep their digest and you do not have to save the images locally and move them to the air-gapped registry before pushing them.

Reviewer: ```suggestion
    If your Platform Administrator Machine has access to the air-gapped container registry mirror, you can use a tool like [`regctl` or `regsync`](https://github.com/regclient/regclient) which can help syncing images between repositories. This way, images will keep their digest and you do not have to save the images locally and move them to the air-gapped registry before pushing them.
```

```yaml
alerts:
  alertsTo: email
  customReceivers:
  - name: 'email'
    email_configs:
    - to: 'admin@example.com'
      from: 'prometheus@example.com'
      require_tls: false
      send_resolved: true
```

## Demo

Following are some screenshots of an air-gapped Compliantkubernetes environment, accessing Service Endpoints on the local air-gapped domain by using a [SOCKS proxy](https://en.wikipedia.org/wiki/SOCKS) in Firefox over an SSH tunnel. Accessing Grafana Dashboards:

![](../img/air-gapped-grafana.png)

Signing in with Dex will redirect to the Dex Service Endpoint for authentication (in this case a static user is used):

![](../img/air-gapped-dex.png)

Checking the endpoint of the [user demo](https://github.com/elastisys/compliantkubernetes/tree/main/user-demo) application:

![](../img/air-gapped-demo.png)

Seeing HTTP metrics for the `/users` endpoint of the user demo in Grafana Dashboards:

![](../img/air-gapped-grafana-demo.png)

Seeing logs from the user demo pods in OpenSearch Dashboards:

> ![](../img/air-gapped-opensearch-demo.png)

Reviewer 3: The demo pictures are very small and hard to actually see what's going on in them on the website.
![image](https://github.com/elastisys/compliantkubernetes/assets/91016401/77f3c240-2169-4ced-92ae-e79203c23efb)

Also, is there some kind of personal data related stuff with you showing your last name on the screenshots? Not sure what was decided on that. @cristiklein will have to elaborate on that.

Reviewer: Let's use `.air-gapped.internal` as suggested by @salehsedghpour : https://www.icann.org/en/public-comment/proceeding/proposed-top-level-domain-string-for-private-use-24-01-2024

It feels more professional and clearly signals "this doesn't work on the Internet".

Author: @AlbinB97 PTAL if the pictures are good now or if they should be zoomed in even more. I changed to the `.air-gapped.internal` domain now as well

Reviewer 3: Looks way better now, great job! :+1:

## References

- https://en.wikipedia.org/wiki/Air_gap_(networking)

> - https://kubespray.io/#/docs/offline-environment

Reviewer 3: These are not actual clickable links

Author: Nice catch, should be fixed now.
